docs/prover/approx/hashing.md

```diff
@@ -0,0 +1,182 @@
+	uint u, v, w;
+	// ...
+		require b1.length < 224;
+		m[b1] = u;
+	// ...
+		// no constraints on b2.length
+		m[b2] = v; 
+	// ...
+		m[b3] = v;
+	    assert(b3.length > 300, "b3 is less than 300 bytes long, unexpectedly")
+	// ...
+}
+```
+
+Let us assume that the `--hashing_length_bound` flag is set to 224 (which corresponds to 7 machine words).
+Then, the first hash operation, triggered by the mapping access `m[b1]`, behaves like the hash of a bounded data chunk. The `--optimstic_hashing` flag has no impact on this hash operation.
+Behavior of the second hash operation, triggered by the mapping access `m[b2]`, depends on whether `--optimistic_hashing` is set. 
+If the `--optimistic_hashing` flag is not set, the violation of an internal assertion will be reported by the prover, stating that an chunk of data is being hashed that may exceed the given bound of 224.
+If the `--optimistic_hashing` flag is set, the prover will internally impose an assumption (like a `require` statement) on `b2` stating that its length cannot exceed 224 bytes.
+The third operation behaves like the second, since also no length constraint on `b3` is made by the program. However, we can see the impact of the `--optimistic_hashing` flag on the `assert` command that follows the hash operation: When the flag is set, the assertion will be shown as not violated even though nothing in the program itself prevents `b3` from being longer than 300 bytes. This is an example of potential unsoundness coming from "optimistic" assumptions.
+(When `--optimistic_hashing` is not set, then we get a violation from any or all assertions, depending on the configuration of Certora Prover.)
+
+
+
+
+
+
+### Examples for Unbounded Hashing
+
+The following collection snippet illustrates the most common use cases for hashing of data that has unbounded length.
+
+
+```solidity
+contract C {
+	mapping(bytes => uint) m; 
+	uint x, y, z, start, len;
+    // ... 
+		m[b] = v
+    // ... 
+		keccak256(abi.encode(x, y, z))
+    // ... 
+		keccak256(abi.encodePacked(x, y, z))
+	// ...
+		assembly {
+			keccak(start, len)
+		}
+	// ...
+}
+
+```
+
+Probably the most common use case is the use of mappings whose keys are an unbounded array (`bytes`, `string`, `uint[]`, etc.); any access to such a mapping induces a hash of the corresponding array whose length is often unknown and unbounded.
+
+Further use cases include direct calls of the Keccak function, either directly on solidity or inside an inline assembly snippet.
+
+Note that Certora Prover's static analysis is aware of the ABI encoder. Thus, in many cases, it can figure out that when `x, y, z` are scalars that `keccak256(abi.encode(x, y, z))` is actually a bounded hash of the form `hash(x, y, z)` as opposed to an unbounded hash of the `bytes` array that is the result of the `encode` function.
+
+
+
+
+## Conclusion
+
+To summarize, Certora Prover handles hashing in a way such that for the vast majority of hashes it will behave as expected. 
```

Comment:
"...in a way that behaves as expected for most hashes."

Reply:
changed